### ISSUE TITLE
Proposal: SiaMux frame batching

### ISSUE THREAD

**Issue by lukechampine** [SiaFoundation/core#21]
SiaMux currently wastes a lot of bandwidth when processing many small writes. Each write is enclosed within a frame, and each frame is padded to the next packet boundary. Assuming a 1500-byte Ethernet MTU, this means that sending a 16-byte identifier incurs 99% overhead. If there is only one write, this is the best we can do. But in the presence of multiple concurrent writes, we can achieve much greater efficiency by batching the writes together.

SiaMux has a flag called `frameContainsAnotherFrame` that seems relevant here. The flag is currently unused, but the idea is that you can store one frame within the padding of another. The original motivation was to enable "covert" streams that are undetectable to traffic sniffers. But combining frames in this way clearly increases efficiency -- and nothing stops us from applying the flag recursively, thus batching many frames into a single packet.

However, this strategy has a significant downside: since frames are only padded to the next packet boundary, the number of additional frames we can fit into this padding is quite limited. The implementation is a little awkward too (I tried), which should not be surprising considering that `frameContainsAnotherFrame` was not designed for this purpose.

In light of this, I propose tweaking the SiaMux protocol to enable first-class support for frame batching. The new primitive will be the *message*, a batch of frames preceded by the following (encrypted) header:

```go
type MessageHeader struct {
    flags  uint16
    length uint32
}
```

This header is followed by an encrypted payload with the given `length`, padded to the next packet boundary, up to a maximum of `MaxMessagePackets` (renamed from `MaxFrameSizePackets`). Inside the decrypted payload are a series of frames, which have the same structure as they do now. Covert streams are implemented essentially the same way as before, without the need for a special flag: if we can add a covert frame to a message without increasing the total number of packets sent, we do so. (We must be careful here, though, because ordering is significant: the covert frame *must be the last frame* added to the message, or else it may affect the number of packets sent.)

The `flags` field is mostly for future-proofing; I suggest using it to signal `idEstablishEncryption`, `idUpdateSettings`, and `idKeepalive` instead of using frames for those messages, but it's not a huge deal either way.

In summary, this change can significantly improve efficiency in the presence of many small concurrent writes, at the cost of 6 additional bytes of overhead. (And since batching will reduce the number of encryption nonces we need, those 6 bytes will "pay for themselves" very quickly.)

Thoughts?

**Comment by ChrisSchinnerl** [comment]
Ok so this boils down to:

1. introduce the concept of a message which contains multiple frames which contains multiple packets
2. move relevant flags from the frame to the message but keep the flags field in the frame as well
3. increase siamux protocol version byte from 1 to 2.

At this point I'd like to mention that we also [tried](https://gitlab.com/NebulousLabs/siamux/-/merge_requests/42) (and then abandoned the MR for a year). As it turns out it's indeed awkward.

So my first thought is that I like the proposal and that it's also a good chance to get rid of unnecessary allocations within the siamux. We might also see a huge improvements in latency for looking up tons of registry entries and sectors in parallel.

**Comment by DavidVorick** [comment]
Why do we need to modify the protocol? Doesn't the `frameContainsAnotherFrame` flag give us plenty of room to do padding already? I believe frames already do everything that you want the messages to do, we should be able to get batching implemented without needing to add more extensions.

From what I remember one of the major challenges was 'how do you know when to batch things?' If you want to batch things together, you need some sort of queue, which means you are introducing latency. We'd need some sense that the connection is already saturated. Or we'd need to be comfortable adding an artificial delay (like 5ms) to our frames so we can see if more frames can be packed together before making the request.

**Comment by lukechampine** [comment]
Frames are only padded to the next packet boundary, so you can only pack one frame into another if it's smaller than (at most) 1440 bytes. Consider sending two 2000-byte frames. With `frameContainsAnotherFrame`, they can't be combined, so you send four total packets. But if you could batch them into one message, you would only need three total packets. (The three packets would also be encrypted together, using one nonce + digest, instead of needing two encrypt calls with two nonces and digests.)

Moreover, though, I think explicitly concatenating frames is just a cleaner design than recursively stuffing extra frames inside each other's padding. `frameContainsAnotherFrame` makes perfect sense if the goal is to support covert streams, but using it to *batch* frames feels like a hack.

Re: "when to batch" -- I figure we can start by batching whenever the `writeLoop` is "not looking," i.e. if it's blocked on a `Write` syscall, or if it's slow to reacquire the mutex because there's a lot of contention. The downside is that, since `Write` generally does not block for longer than a few microseconds, you need a *lot* of write activity for your frames to be batched together. (We should benchmark this, of course.) I'm wary of introducing artificial delays, but they're probably unavoidable if we want to batch more aggressively. If we *do* add them, we'll certainly want to add an escape hatch as well, so that we can minimize latency for important messages.

I gotta say, this is starting to feel uncomfortably similar to `TCP_NODELAY`. We should try to avoid reinventing TCP -- until we're ready to *actually* reinvent TCP, that is. 😅

**Comment by DavidVorick** [comment]
>  Frames are only padded to the next packet boundary, so you can only pack one frame into another if it's smaller than (at most) 1440 bytes. Consider sending two 2000-byte frames.

If you know in advance that you are packing the frames together, you can just make the first frame 3 packets. The frame decoder doesn't care that you used extra packets, that's not one of the validation steps.

> Moreover, though, I think explicitly concatenating frames is just a cleaner design than recursively stuffing extra frames inside each other's padding. frameContainsAnotherFrame makes perfect sense if the goal is to support covert streams, but using it to batch frames feels like a hack.

The design actually went the other way around. `frameContainsAnotherFrame` was invented as a method to push frames together without the overhead of building a declaration on top. Instead of 3 bytes, it's one bit. Also, covert streams can fit into any design that is fundamentally padding every message to an integer number of packets.

> The downside is that, since Write generally does not block for longer than a few microseconds, you need a lot of write activity for your frames to be batched together.

That's the challenge, `Write` is not necessarily going to block even if your connection is already saturated and has built up a buffer. I'm not sure how to get around that. The actual worker code can make use of this as well if it knows the network is blocked (because it can batch programs together to save on MDM overhead), but for similar reasons as siamux it doesn't really have a way to tell that there's a good opportunity to wait for stuff to be batched.

**Comment by lukechampine** [comment]
> If you know in advance that you are packing the frames together, you can just make the first frame 3 packets. The frame decoder doesn't care that you used extra packets, that's not one of the validation steps.

I don't see how this would work. The decoder starts by reading the frame header, which contains a length. It then rounds that length up to the next packet boundary, and reads that many bytes. Then it trims the result to `buf[:length]` to get the intended payload.

If we append more data without modifying `length`, then the decoder will read the same number of packets as before. If we *do* increase `length`, then we end up reading a payload that is the concatenation of multiple frames -- but how do we know where the frame boundaries are? The single `frameContainsAnotherFrame` bit doesn't give us enough information to determine where one frame ends and the next one begins.

**Comment by DavidVorick** [comment]
> I don't see how this would work. The decoder starts by reading the frame header, which contains a length. It then rounds that length up to the next packet boundary, and reads that many bytes. Then it trims the result to buf[:length] to get the intended payload.

You can check if `frameContainsAnotherFrame` is set, and then seek ahead to that frame to check the size of that frame, and continue until you get through the whole message.

**Comment by lukechampine** [comment]
Ah, ok. But now you need to make two `Read` calls for every frame, on top of needing two decryptions. If you're batching many small messages, that overhead adds up quickly.

Here's a compromise: we redefine `frameContainsAnotherFrame` as `superFrame`. When the `superFrame` is set, it basically means that the recipient should treat the payload as a contiguous sequence of frames. With this, we sorta get the best of both worlds: batched frames only require one `Read` and decrypt, *and* we don't have to add any overhead to non-batched frames. Arguably we don't even need to bump the protocol version, since `frameContainsAnotherFrame` was never precisely specified or implemented.

**Comment by DavidVorick** [comment]
> Arguably we don't even need to bump the protocol version, since frameContainsAnotherFrame was never precisely specified or implemented.

We need to bump the protocol version either way, current users of siamux don't have any software at all to decode FCAF messages. So we need a bump if we start doing any form of this because we need to know when the other guy can't receive the batched frames.

> When the superFrame is set, it means that the recipient should treat the payload as a contiguous sequence of frames. With this, we sorta get the best of both worlds: batched frames only require one Read and decrypt, and we don't have to add any overhead to non-batched frames.

I don't understand exactly how this works, can you explain it more?

**Comment by lukechampine** [comment]
> We need to bump the protocol version either way

Oh right, duh.

> I don't understand exactly how this works, can you explain it more?

If there's only one frame to send, just send it. If there are multiple frames, then send:
```
[superFrame header] [header 0] [payload 0] [header 1] [payload 1] ...
```
Where `superFrame header` has the `superFrame` flag set, and has a `length` equal to the length of the concatenated headers+payloads.

**Comment by lukechampine** [comment]
Update: we've agreed that adding a single flag might be all that's necessary. If `flagNoPadding` is set, then the decoder should expect another frame immediately after the current one.

The benefit of this approach is that we don't need to introduce a `message` or `superFrame` construct, and we still get the efficiency benefits of eliminating padding. Another benefit is that we can process messages as soon as they arrive, reducing latency; this is a prominent disadvantage of my original proposal, where frame 0 can only be processed after reading+decrypting the entire message. We've agreed that the overhead of encrypting each frame separately is an acceptable price to pay for lower latency (although we should definitely do some exploratory benchmarking before making the final call here).

**Comment by lukechampine** [comment]
I ran into some difficulties with `flagNoPadding` that eventually led me to consider a different approach.

The new approach, it a nutshell, is to handle encryption at the packet level. On the wire, we send a sequence of packets. Every packet is the same size and every packet is encrypted. Padding, if necessary, is added prior to encryption. By decrypting the sequence of packets, you get a stream of frame data + padding. As usual, the frame data is decoded and the padding is discarded.

Visually:
```
Old protocol:
[ encrypted header ] [ encrypted payload ] [ encrypted header ] [ encrypted payload ]  [ padding ] 
[__________ packet 0 __________] [__________ packet 1 __________] [__________ packet 2 __________]

New protocol:
[ header ] [ payload ] [ header ] [ payload ] [     padding     ] 
[_____ encrypted packet 0 _____] [_____ encrypted packet 1 _____]
```

This has a few implications:
- There is less time+space overhead for small frames, because multiple frames can be covered by one encryption. (Note that this does *not* increase latency, because you never need to wait for more packets to arrive before processing the first frame.)
- There is more time+space overhead for large frames, because they are split into packets which are encrypted separately. (Even with this extra overhead, though my benchmarks are hitting ~500 MB/s throughput and >300,000 streams per second.)
- There is no longer a need for a `MaxFrameSizePackets` setting, since frames are split into packets.
- `frameHeader.length` can be a `uint16`, since it should never exceed the size of 1 packet.
- You can read a full frame (or multiple frames) in a single `Read` syscall, instead of reading once to get the header and again to get the payload.
- Encryption/decryption can be implemented as wrappers around `net.Conn`, allowing the main mux logic to deal exclusively with plaintext. I imagine this would work nicely with UDP as well, where everything is packet-oriented anyway.

Of course, we still have to handle the padding somehow. The mux will now see an undifferentiated stream of bytes that contain fixed-length headers, variable-length payloads, and variable-length padding; how do we figure out where the boundaries are? My solution is to steal 1 bit from the `frameHeader.id`. If you see a `1`, that means it's a frame header (followed by a payload); if you see a `0`, it's padding, so seek forward until you hit the next frame header. I was originally worried that naively seeking like this would kill performance, but profiling didn't indicate much impact at all.

Lastly, covert streams. The simplest approach would be insert covert frames whenever the padding is large enough to accommodate them. However, we can do better by stealing another bit from `frameHeader.id`. If this bit is set, the padding should be treated as a covert data stream, i.e. a header + payload that is split across (potentially) many runs of padding. The upside is that you only need one header for the entire covert payload, instead of splitting the payload into a bunch of smaller frames so that they can fit inside the available padding.

Thoughts?
